### ISSUE TITLE
Zero comparison - breaking change?

### ISSUE THREAD

**Issue by james-trayford** [yt-project/unyt#369]
I've been looking into `unyt` as a dependency of another package, and noticed that the change in behaviour in `2.9.4` from `2.9.2` (latest previous version tested _**Edit**: behaviour arises in v2.9.3, see below_) regarding zero comparison. This breaks a number of aspects of the  dependent code.

Illustrating the issue, here is the `2.9.4` behaviour:

```
>>> unyt.__version__
'v2.9.4'
>>> a = unyt.unyt_array(np.zeros(5, dtype=np.int8), dtype=np.int8)
>>> for i in a:
...     print(i == 0)
... 
False
False
False
False
False
>>> 
```

Then with `unyt` `2.9.2` (latest previous version tested):

```
>>> unyt.__version__
'v2.9.2'
>>> a = unyt.unyt_array(np.zeros(5, dtype=np.int8), dtype=np.int8)
>>> for i in a:
...     print(i == 0)
... 
True
True
True
True
True
>>> 
```

should this be regarded a breaking change, i.e. correspond to a new major version number?

**Comment by MatthieuSchaller** [comment]
FWIW, I reckon that both behaviours are wrong. `unyt` ought to throw an exception if users are comparing a dimension-full array to any dimension-less constant (including 0). Not `True` or `False`. 

Though, code (like ours) built on the assumption it returns `True` is broken by the change.

**Comment by ngoldbaum** [comment]
This was likely an unintentional change. Ping @neutrinoceros.

**Comment by james-trayford** [comment]
> FWIW, I reckon that both behaviours are wrong. unyt ought to throw an exception if users are comparing a dimension-full array to any dimension-less constant (including 0). Not True or False.
>
> Though, code (like ours) built on the assumption it returns True is broken by the change.

Generally agree, I suppose this gets a bit more of a grey area when dealing with `dimensionless` arrays specifically (i.e. when the unit in question is `[dimensionless]`) though.

**Comment by neutrinoceros** [comment]
unyt v2.9.3 is the first one with the current behaviour, not 2.9.4
The change happened as a side effect of https://github.com/yt-project/unyt/pull/333, and as far as I can tell/remember, it was indeed not intended.
I'll give it a closer look and try to understand what's happening here.

**Comment by ngoldbaum** [comment]
> FWIW, I reckon that both behaviours are wrong. unyt ought to throw an exception if users are comparing a dimension-full array to any dimension-less constant (including 0). Not True or False.

For what it's worth, the behavior of ignoring the units when comparing with a quantity that has a value of zero was added intentionally. In general we try to err on the side of not raising errors when they aren't strictly needed. See https://github.com/yt-project/unyt/commit/194440ec74df943856bcad3307a2f43501e60606. It looks like the main reason this got added was to more naturally support the `sum` builtin, which has a default `start` value of 0.

**Comment by neutrinoceros** [comment]
We have a fix #371 
I'll make a new bugfix release with it. Thank you for reporting !

**Comment by james-trayford** [comment]
Thanks for the rapid fix!

**Comment by neutrinoceros** [comment]
This is fixed in unyt 2.9.5
